### DISCUSSION TITLE
OStream mutex addition breaks managed code (C++/CLI) compilations.

### DISCUSSION THREAD

**Discussion by BStringhamVRSK** [google/filament#5329] [Q&A]
@pixelflinger @prideout @bejado 
We are using Filament in a C# managed code environment.  **std::mutex** is not supported when C++ code is compiled with the /clr flag.  Because the std::mutex declaration appears in the include header file, it breaks our builds.

There are two ways forward that I can see:

1.  Move the mutex into ostream.cpp as a global.  That means all ostream instances will share the same mutex.  PROS:  This is probably the simplest change.  CONS:  This could have some impact on performance, but I don't know if it's a meaningful difference.
2. Create an **interface** for ostream.  That would require also creating a "factory" function to create ostream instances, rather than using the new operator.  PROS:  ostream would continue to work as before.  CONS:  Breaks any code currently using "new ostream()"

Any other thoughts or suggestions on this?  
I can do the programming changes, but I'd like to hear the code owners' opinions before diving in.

**Comment by romainguy** [comment]
Thanks for catching this, in general we strive to avoid STL dependencies in our public headers. I wouldn't worry about performance since our ostreams are only used for debugging. They are not relied upon to render frames.

@pixelflinger to confirm

**Comment by pixelflinger** [comment]
best would be to file an issue so this stays on my radar.

**Reply by BStringhamVRSK** [reply]
Will do.
